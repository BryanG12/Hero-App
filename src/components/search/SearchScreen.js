import React, { useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm'
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';

export const SearchScreen = () => {


  const navigate = useNavigate();
  const location = useLocation ();

  const { q= ''} = queryString.parse(location.search);

  // useMemo

  const [ formValues, handleInputChange ] = useForm({
    searchText:q
  });
  
  const {searchText} = formValues;
  
  const heroesFileted =  useMemo( () =>  getHeroesByName(q),[q]);
  
  const handleSearch= ( e )=>{
    e.preventDefault();
    navigate(`?q=${searchText}`);
  }


  return (
    <>
      <h1>Búsqueda</h1>
      <hr/>
      <div className='row'>
        <div className='col-5'>
          <h4> Buscar</h4>
          <hr/>

          <form onSubmit={handleSearch}>
            <input
              type='text'
              placeholder='Buscar un héroe'
              className='form-control'
              name='searchText'
              autoComplete='off'
              value={ searchText }
              onChange={handleInputChange }
            />
            <div className='d-grid gap-2'>
              <button
                className='mt-3 btn btn-warning btn-block'
                type='submit'
                
              >
                Buscar...

              </button>
            </div>
          </form>
        </div>

        <div className='col'>
          <h4>Resultados</h4>
          <hr/>

          {
            (q === '')
              ? <div className='alert alert-info'> Buscar un héroe</div>
              : (heroesFileted.length === 0 ) 
                && <div className='alert alert-danger'> No hay resultados: {q} </div>
          }



          {
            heroesFileted.map( hero => (
              <HeroCard
                key={hero.id}
                {...hero} 
              />
            ))
          }
        </div>

      </div>
    </>
  )
}
