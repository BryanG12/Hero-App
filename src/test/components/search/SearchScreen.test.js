import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';


describe('pruebas en <SearchScreen />', () => { 

  test('should de mostrar correctamente con valores por defecto', () => { 

    
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen /> 

      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un hÃ©roe');
  })

  test('should de mostrar a Batman y el input con el valor del queryString', () => { 

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen /> 

      </MemoryRouter>
    );
    
    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper).toMatchSnapshot();

  })

  test('should de mostrar un error si no se encuentra el heroe', () => { 

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']} >
        <SearchScreen />
      </MemoryRouter>
    );


    expect(wrapper.find('.alert-danger').exists()).toBe(true);
    expect(wrapper.find('.alert-danger').text().trim()).toBe('No hay resultados: batman123');


  })
});