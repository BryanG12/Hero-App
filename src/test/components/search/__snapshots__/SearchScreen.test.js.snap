// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <SearchScreen /> should de mostrar a Batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Búsqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value="batman"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="mt-3 btn btn-warning btn-block"
            type="submit"
          >
            Buscar...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                   
                  Batman
                   
                </h5>
                <p
                  className="card-text"
                >
                   
                  Bruce Wayne
                   
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                     
                    Detective Comics #27
                     
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`pruebas en <SearchScreen /> should de mostrar correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Búsqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="mt-3 btn btn-warning btn-block"
            type="submit"
          >
            Buscar...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
         Buscar un héroe
      </div>
    </div>
  </div>,
]
`;
